File: info/1. Алгоритмы/3. Поиск/6. Поиск Фибоначчи/code sample/fibonacci_search.py
class FibonacciSearch:
    def __init__(self):
        self.i = 0
        self.p = 0
        self.q = 0
        self.stop = False

    def get_fibonacci_number(self, k):
        first_number = 0
        second_number = 1
        n = 0
        while n < k:
            temp_number = second_number
            second_number = first_number + second_number
            first_number = temp_number
            n = n + 1
        return first_number

    def start_init(self, sequince):
        self.stop = False
        k = 0
        n = len(sequince)
        while (self.get_fibonacci_number(k+1) < len(sequince)):
            k = k+1
        m = self.get_fibonacci_number(k+1)-(n+1)
        self.i = self.get_fibonacci_number(k) - m
        self.p = self.get_fibonacci_number(k-1)
        self.q = self.get_fibonacci_number(k-2)

    def up_index(self):
        if self.p == 1:
            self.stop = True
        self.i = self.i + self.q
        self.p = self.p - self.q
        self.q = self.q - self.p

    def down_index(self):
        if self.q == 0:
            self.stop = True
        self.i = self.i-self.q
        temp = self.q
        self.q = self.p - self.q
        self.p = temp

    def search(self, sequince, element):
        self.start_init(sequince)
        result_index = -1
        while not self.stop:
            if self.i < 0:
                self.up_index()
            elif self.i >= len(sequince):
                self.down_index()
            elif sequince[self.i] == element:
                result_index = self.i
                break
            elif element < sequince[self.i]:
                self.down_index()
            elif element > sequince[self.i]:
                self.up_index()
        return result_index


sequince = [-2, 0, 3, 5, 7, 9, 11, 15, 18, 21]


fs = FibonacciSearch()

element = 7

print(fs.search(sequince, element))

File: info/1. Алгоритмы/3. Поиск/6. Поиск Фибоначчи/code sample/FibonacciSearch.java

public class FibonacciSearch {
    private int i;
    private int p;
    private int q;
    private boolean stop = false;

    public FibonacciSearch() {

    }

    private void init(int[] sequince) {
        stop = false;
        int k = 0;
        int n = sequince.length;
        for (; getFibonacciNumber(k + 1) < n + 1;) {
            k += 1;
        }
        int m = (int) (getFibonacciNumber(k + 1) - (n + 1));
        i = (int) (getFibonacciNumber(k) - m);
        p = (int) getFibonacciNumber(k - 1);
        q = (int) getFibonacciNumber(k - 2);
    }

    public long getFibonacciNumber(int k) {
        long firstNumber = 0;
        long secondNumber = 1;
        for (int i = 0; i < k; i++) {
            long temp = secondNumber;
            secondNumber += firstNumber;
            firstNumber = temp;
        }
        return firstNumber;
    }

    private void upIndex() {
        if (p == 1)
            stop = true;
        i = i + q;
        p = p - q;
        q = q - p;
    }

    private void downIndex() {
        if (q == 0)
            stop = true;
        i = i - q;
        int temp = q;
        q = p - q;
        p = temp;
    }

    public int search(int[] sequince, int element) {
        init(sequince);
        int n = sequince.length;
        int resultIndex = -1;
        for (; !stop;) {
            if (i < 0) {
                upIndex();
            } else if (i >= n) {
                downIndex();
            } else if (sequince[i] == element) {
                resultIndex = i;
                break;
            } else if (element < sequince[i]) {
                downIndex();
            } else if (element > sequince[i]) {
                upIndex();
            }
        }
        return resultIndex;
    }

}
